torch/lib/c10d/ProcessGroupWrapper.cpp

```diff
@@ -0,0 +1,314 @@
+    std::vector<at::Tensor>& outputTensors,
+    std::vector<std::vector<at::Tensor>>& inputTensors,
+    const ScatterOptions& opts) {
+  runCollectiveChecks(OpType::SCATTER, outputTensors);
+  return pg_->scatter(outputTensors, inputTensors, opts);
+}
+
+c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupWrapper::reduce_scatter(
+    std::vector<at::Tensor>& outputTensors,
+    std::vector<std::vector<at::Tensor>>& inputTensors,
+    const ReduceScatterOptions& opts) {
+  runCollectiveChecks(OpType::REDUCE_SCATTER, outputTensors);
+  return pg_->reduce_scatter(outputTensors, inputTensors, opts);
+}
+
+c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupWrapper::alltoall_base(
+    at::Tensor& outputTensor,
+    at::Tensor& inputTensor,
+    std::vector<int64_t>& outputSplitSizes,
+    std::vector<int64_t>& inputSplitSizes,
+    const AllToAllOptions& opts) {
+  // alltoall supports uneven split, so don't enforce shape checking.
+  runCollectiveChecks(OpType::ALLTOALL_BASE, {});
+  return pg_->alltoall_base(
+      outputTensor, inputTensor, outputSplitSizes, inputSplitSizes, opts);
+}
+
+c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupWrapper::alltoall(
+    std::vector<at::Tensor>& outputTensors,
+    std::vector<at::Tensor>& inputTensors,
+    const AllToAllOptions& opts) {
+  // alltoall supports uneven split, so don't enforce shape checking.
+  runCollectiveChecks(OpType::ALLTOALL, {});
+  return pg_->alltoall(outputTensors, inputTensors, opts);
+}
+
+void ProcessGroupWrapper::monitoredBarrier(
+    const BarrierOptions& opts,
+    bool waitAllRanks) {
+  return pg_->monitoredBarrier(opts, waitAllRanks);
+}
+
+void ProcessGroupWrapper::setSequenceNumberForGroup() {
+  // Set underlying pg's sequence number if it is not set.
+  if (pg_->getSequenceNumberForGroup() == 0) {
+    // Set the sequence number for the underlying process group.
+    pg_->setSequenceNumberForGroup();
+  }
+}
+
+uint64_t ProcessGroupWrapper::getSequenceNumberForGroup() {
+  return pg_->getSequenceNumberForGroup();
+}
+
+c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupWrapper::send(
+    std::vector<at::Tensor>& tensors,
+    int dstRank,
+    int tag) {
+  return pg_->send(tensors, dstRank, tag);
+}
+
+c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupWrapper::recv(
+    std::vector<at::Tensor>& tensors,
+    int srcRank,
+    int tag) {
+  return pg_->recv(tensors, srcRank, tag);
+}
+
+c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupWrapper::recvAnysource(
+    std::vector<at::Tensor>& tensors,
+    int tag) {
+  return pg_->recvAnysource(tensors, tag);
+}
+
+c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupWrapper::barrier(
+    const BarrierOptions& opts) {
+  runCollectiveChecks(OpType::BARRIER, {});
+  return pg_->barrier(opts);
+}
+
+void ProcessGroupWrapper::runCollectiveChecks(
+    OpType op_type,
+    const std::vector<at::Tensor>& tensors) const {
+  auto finger_print = CollectiveFingerPrint(op_type, tensors);
```

Comment:
are you planning to run monitored_barrier() as well before collectives?

Reply:
Yes we can add this, thanks.